clone/internal/download/batch.go

```diff
@@ -0,0 +1,108 @@
+// Copyright 2021 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//    https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// download contains classes for downloading data from logs.
+package download
+
+import (
+	"fmt"
+	"time"
+
+	backoff "github.com/cenkalti/backoff/v4"
+	"github.com/golang/glog"
+)
+
+// LeafFetcher is an interface to be implemented to provide a mechanism to fetch a range of leaves.
+// The leaves returned should be those in the range [start, end).
+type LeafFetcher interface {
+	Batch(start, end uint64) ([][]byte, error)
+}
+
+// Bulk keeps downloading leaves starting from `first`, using the given leaf fetcher.
+// The number of workers and the batch size to use for each of the fetch requests are also specified.
+// The resulting leaves are returned in order over `leafChan`, and any terminal errors are returned via `errc`.
+// Internally this uses exponential backoff on the workers to download as fast as possible, but no faster.
+func Bulk(first uint64, lf LeafFetcher, workers, batchSize uint, leafChan chan<- []byte, errc chan<- error) {
+	// Each worker gets its own unbuffered channel to make sure it can only be at most one ahead.
+	// This prevents lots of wasted work happening if one shard gets stuck.
+	rangeChans := make([]chan leafRange, workers)
+
+	increment := workers * batchSize
+	for i := uint(0); i < workers; i++ {
+		rangeChans[i] = make(chan leafRange)
+		start := first + uint64(i*batchSize)
+		go fetchWorker{
+			label:     fmt.Sprintf("worker %d", i),
+			start:     start,
+			end:       start + uint64(batchSize),
```

Comment:
naming `start` (or `first`) and `count` would help avoid any ambiguity about the in/exclusivity in the ranges.